library/core/src/iter/adapters/zip.rs

```diff
@@ -416,6 +416,10 @@ impl<A: Debug + TrustedRandomAccess, B: Debug + TrustedRandomAccess> ZipFmt<A, B
 
 /// An iterator whose items are random-accessible efficiently
 ///
+/// Iterators that implement this trait should also implement TrustedLen which
+/// allows specialization to disambiguate overlaps with a `TrustedLen + TrustedRandomAccess`
+/// bound.
```

Comment:
Hm, if this isn't the case, what happens? Is it just a loss of optimization potential? Compile error?

I'm not entirely sure, but it seems like TrustedRandomAccess does not imply TrustedLen in all cases -- particularly, there may be a leading piece of the iterator which *is* TrustedRandomAccess and the tail is not, so TrustedLen cannot be implemented but TrustedRandomAccess seems like it can be (as self.size() takes the lower of the two size hints). So this comment I think could use elaboration.

Reply:
Hm, I guess I must be missing the safety requirement that requires an exact length...

Reply:
It's the first line in the Safety section:

>  The iterator's `size_hint` must be exact and cheap to call.

Reply:
Ah, I guess I don't see that as meaning "trusted" (since ExactSizeIterator doesn't imply this). It seems odd that we don't have a Self: TrustedLen bound on TrustedRandomAccess then.

Reply:
They have different origins. `TrustedRandomAccess` initially only existed to optimize `Zip`, `TrustedLen` existed to optimize `FromIterator`. Since then their uses have grown and started to partially overlap.